### ISSUE TITLE
"Terrapin" MitM attack / CVE-2023-48795

### ISSUE THREAD

**Issue by bitprophet** [paramiko/paramiko#2337]
This is just a tracking ticket to note that we are aware of the issue described here: https://terrapin-attack.com/ - and have been working with the researchers to release a new Paramiko version that mitigates the described attack pattern.

We expect to put out Paramiko 3.4 within the next day or so. The work in progress is here: https://github.com/paramiko/paramiko/tree/cve-2023-48795

**Comment by bitprophet** [comment]
CI is green now that I fixed some speling and inserted the final OpenSSH version number in our changelog. Waiting on a final thumbs up from the researchers re: additional fixes I put in yesterday, and will then plop this out on PyPI.

**Comment by bitprophet** [comment]
3.4 is on PyPI! https://www.paramiko.org/changelog.html#3.4.0

**Comment by anarcat** [comment]
@bitprophet could you clarify exactly what is the patch here? https://github.com/paramiko/paramiko/commit/96db1e2be856eac66631761bae41167a1ebd2b4e and https://github.com/paramiko/paramiko/commit/33508c920309860c4a775be70f209c2a400e18ec? 

i believe debian will need to backport the fixes to 2.7.2 and 2.4.2, your input on that would be tremendously useful...

**Comment by carnil** [comment]
Not upstream, but there is above mentioned a branch where the fixes were developed, so all what is between 2566d2d5d94d331041d8ca3ab1d72ada93e642e2...3e4bdf998f5b1508322234a527e8fa432220368b is needed for the fixes?

**Comment by bitprophet** [comment]
Yup, @carnil's got it, that diff notation appears to match to the branch linked in the OP.

(I'm not sure why GH doesn't make it easier to examine branches post-merge. When I use tools like `git log` or `lazygit` I can see eg where the CVE branch branches off of the 3.3 branch, for example, which would have been == `main` at time of development.)

**Comment by bitprophet** [comment]
@anarcat ☝🏻 🙏🏻

**Comment by coldtobi** [comment]
Hi @bitprophet ,

I'm currently trying to fix paramiko for Debian buster (LTS). The version of paramiko in Debian buster is 2.4.2 -  and I'd like to socialite your expertise to help me if my thoughts are correct.  

While backpacking the patches, I realized that 2.4.2. might not be vulnerable to terrapin at all, as it seems not to support certain features required to exploit the vulnerability: 
If I understood terrapin correctly it requires ChaCha20-Poly1305 or CBC-EtM support, and I think 2.4.2 does not have those.  (I also think that 2.4.2 does not support EXT_INFO, but not sure on that.)

When using the terrapin [Vulnerability Scanner ](https://github.com/RUB-NDS/Terrapin-Scanner/releases/tag/v1.1.0), the tool seems to support the idea that this version is not vulnerable: (this is without the backported patches)

```
tobi@debian10-lts:~/build$ ./Terrapin_Scanner_Linux_amd64 -connect localhost:2200
================================================================================
==================================== Report ====================================
================================================================================

Remote Banner: SSH-2.0-paramiko_2.4.2

ChaCha20-Poly1305 support:   false
CBC-EtM support:             false

Strict key exchange support: false

The scanned peer supports Terrapin mitigations and can establish
connections that are NOT VULNERABLE to Terrapin. Glad to see this.
For strict key exchange to take effect, both peers must support it.

Note: This tool is provided as is, with no warranty whatsoever. It determines
      the vulnerability of a peer by checking the supported algorithms and
      support for strict key exchange. It may falsely claim a peer to be
      vulnerable if the vendor supports countermeasures other than strict key
      exchange.
```

Please correct any flaws in my thinking ;-)

Thanks a lot,
tobi

**Comment by bitprophet** [comment]
Hi @coldtobi,

It looks like the scanner is accurate; I just looked it up and we only added ETM MAC modes in Paramiko 2.5. (and ChaCha20 is still not implemented, fwiw).

And even Paramiko 2.5-2.8 are likely not super vulnerable, as EXT_INFO was only implemented in 2.9, and my understanding is that one of the main practical attacks with Terrapin is to force one end of the connection to disregard EXT_INFO.
